apps/nextjs/src/trpc/server.ts

```diff
@@ -1,40 +1,73 @@
-"use server";
-
-import { headers } from "next/headers";
-import { loggerLink } from "@trpc/client";
-import { experimental_createTRPCNextAppDirServer } from "@trpc/next/app-dir/server";
+import { cache } from "react";
+import { cookies, headers } from "next/headers";
+import { NextRequest } from "next/server";
+import { getAuth } from "@clerk/nextjs/server";
+import { createTRPCClient, loggerLink, TRPCClientError } from "@trpc/client";
+import { callProcedure } from "@trpc/server";
+import { observable } from "@trpc/server/observable";
+import type { TRPCErrorResponse } from "@trpc/server/rpc";
 
+import { appRouter, createTRPCContext } from "@acme/api";
 import type { AppRouter } from "@acme/api";
 
 import { endingLink, transformer } from "./shared";
 
-export const api = experimental_createTRPCNextAppDirServer<AppRouter>({
-  config() {
-    return {
-      transformer,
-      links: [
-        loggerLink({
-          enabled: (opts) =>
-            process.env.NODE_ENV === "development" ||
-            (opts.direction === "down" && opts.result instanceof Error),
-        }),
-        endingLink({
-          headers: () => {
-            const h = new Map(headers());
-            h.delete("connection");
-            h.delete("transfer-encoding");
-            h.set("x-trpc-source", "server");
-            return Object.fromEntries(h.entries());
-          },
+const createContext = cache(() => {
+  return createTRPCContext({
+    headers: new Headers({
+      cookie: cookies().toString(),
+      "x-trpc-source": "server",
+    }),
+    auth: getAuth(
+      new NextRequest("https://notused.com", { headers: headers() }),
+    ),
+  });
+});
+
+export const api = createTRPCClient<AppRouter>({
+  transformer: transformer,
+  links: [
+    loggerLink({
+      enabled: (opts) =>
+        process.env.NODE_ENV === "development" ||
+        (opts.direction === "down" && opts.result instanceof Error),
+    }),
+    endingLink({
+      headers: () => {
+        const h = new Map(headers());
+        h.delete("connection");
+        h.delete("transfer-encoding");
+        h.set("x-trpc-source", "server");
+        return Object.fromEntries(h.entries());
+      },
+    }),
+    /**
+     * Custom RSC link that lets us invoke procedures without using http requests. Since Server
+     * Components always run on the server, we can just call the procedure as a function.
+     */
+    () =>
+      ({ op }) =>
+        observable((observer) => {
```

Comment:
I'm not sure whether this is the correct way, eslint is spouting errors for createContext

Reply:
we wanted to preserver http calls though?

Reply:
fixed